reliability-engineering/terraform/modules/concourse-worker-pool/files/worker-init.sh

```diff
@@ -63,14 +63,56 @@ LimitNOFILE=infinity
 WantedBy=multi-user.target
 EOF
 
+# adapted from
+# https://aws.amazon.com/blogs/compute/best-practices-for-handling-ec2-spot-instance-interruptions/
+# 'EOF' in single quotes prevents varible interpolation; this is important
+cat <<'EOF' > /usr/local/bin/check-spot-interruption
+#!/bin/bash
+set -ue
+TOKEN=`curl -s -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 1800"`
+
+while sleep 5; do
+
+    HTTP_CODE=$(curl -H "X-aws-ec2-metadata-token: $TOKEN" -s -w %%{http_code} -o /dev/null http://169.254.169.254/latest/meta-data/spot/instance-action)
+
+    if [[ "$HTTP_CODE" -eq 401 ]] ; then
+        echo 'Refreshing Authentication Token'
+        TOKEN=`curl -s -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 1800"`
+    elif [[ "$HTTP_CODE" -eq 200 ]] ; then
+        echo 'Interrupted: retiring concourse-worker'
+        systemd kill -s SIGUSR2 concourse-worker
+        exit 0
+    elif [[ "$HTTP_CODE" -eq 404 ]] ; then
+        echo 'Not Interrupted'
+    else
+        echo "Unexpected http code $HTTP_CODE; exiting"
+        exit 1
+    fi
+done
+EOF
+chmod +x /usr/local/bin/check-spot-interruption
+
+cat <<EOF > /etc/systemd/system/check-spot-interruption.service
```

Comment:
non-blocking: I feel like this could have been a systemd timer ... and the loop in the script could have been avoided ... but its simple enough that I don't think it really matters

Reply:
I did consider this, but the AWS blog post provided a script and I'm not confident enough to make my own timer from scratch :) also I don't know if systemd timers work at 5 second resolution.